src/Components/Endpoints/src/Discovery/ComponentCollectionBuilder.cs

```diff
@@ -0,0 +1,96 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.AspNetCore.Components;
+
+/// <summary>
+/// Represents the list of components defined in a <see cref="ComponentApplicationBuilder"/>
+/// before the configuration has finalized.
+/// </summary>
+internal class ComponentCollectionBuilder
+{
+    private readonly List<ComponentBuilder> _components = new();
+
+    internal void Combine(ComponentCollectionBuilder components)
+    {
+        for (var i = 0; i < components._components.Count; i++)
+        {
+            var pageToAdd = components._components[i];
+            var found = false;
+            for (var j = _components.Count - 1; j > 0; j--)
+            {
+                var page = _components[j];
+                if (page.Equals(pageToAdd))
+                {
+                    found = true;
+                    break;
+                }
+            }
```

Comment:
Is it worth storing the items in a dictionary so the collections can be merged without the O(N^2) we have here? I don't know if it would make a difference, but it's definitely believable that people could have component libraries with 1000 components in them.

Reply:
@Eilon asked this same question years ago when we built application parts for MVC, and the answer is that based on measurements that we did at the time, it does not have any visible impact even with a large number of items. (when we tried this in MVC, I think I went up to 10K)

Reply:
That seems truly remarkable - it suggests we could do 100 million loop iterations (10k^2) in the same time as 10k dictionary reads. Are you still able to reproduce those results? Do they apply in this case? Maybe there was something odd going on like the dictionary reads did some very expensive hashing.

I was surprised enough to try this out and got very different results. Perhaps this microbenchmarking isn't a fair comparison though - would be interested to know what you think. I used [this code](https://gist.github.com/SteveSandersonMS/43d59c46a9073cd327ef0bfa86704e97), which compares the nested loop to the dictionary, with results:

```
list took 893ms
dict took 1ms
```

This is merging two collections of size 5k to make one of size 10k.

Reply:
It's clear that the dictionary will be faster than the list, the question is whether it matters. Do we expect having 5k components a common scenario? Even in this case, is less than 1 second not good enough?

Reply:
I've changed this to make it a dictionary. And add some internal structure.